In [1]:
import pandas as pd
import numpy as np
import os

### Etapa 1 - Criação do dataframe com dados populacionais por ano dos municípios brasileiros para servir de base de cáclulo das variáveis [QUANTIDADE DE PROCEDIMENTOS POR 10 MIL HAB] E [VALORES DE PROCEDIEMNTOS POR 10 MIL HAB]

In [2]:
# Criando função para concatenar poulação dos municípios por ano, entre 2010 e 2022, a partir dos dados em arquivo csv na pasta ./data/demografico/ para uso em cálculo posterior
def load_population_data(folder_path):
    dfs = []
    for filename in os.listdir(folder_path):
        if filename.endswith('.csv'):
            year = int(filename.split('.')[0])
            df = pd.read_csv(os.path.join(folder_path, filename), sep=';', names=['codigo', 'pop'], dtype={'codigo': str})
            df.set_index('codigo', inplace=True)
            df = df[~df.index.duplicated(keep='first')] # remove possíveis linhas duplicadas
            df.rename(columns={'pop': str(year)}, inplace=True)
            dfs.append(df.iloc[1:, :]) # ignora a primeira linha
    merged_df = pd.concat(dfs, axis=1)
    merged_df.fillna(0, inplace=True) # preenche valores nulos com 0
    return merged_df


In [3]:
# Aplicando a função fornecendo a pasta com dados como parâmetro
df_pop = load_population_data('E:/OneDrive/Documentos/GitHub/iris/data/demografico')


In [4]:
# checando o dataframe e o resultado da função de concatenação de arquivos csv
df_pop

2010     2011     2012     2013     2014     2015     2016  \
codigo                                                                   
1100015    24392    24228    24069    25728    25652    25578    25506   
1100023    90353    91570    92747   101269   102860   104401   105896   
1100031     6313     6221     6132     6495     6424     6355     6289   
1100049    78574    78958    79330    85863    86556    87226    87877   
1100056    17029    16939    16852    18041    18013    17986    17959   
...          ...      ...      ...      ...      ...      ...      ...   
5222005    12548    12644    12737    13227    13343    13456    13567   
5222054     7371     7475     7576     7933     8053     8171     8286   
5222203     4735     4846     4954     5246     5371     5495     5615   
5222302     5145     5195     5244     5460     5520     5578     5635   
5300108  2570160  2609997  2648532  2789761  2852372  2914830  2977216   

            2017     2018     2019     2020     2021     2022  
codigo                                                         
1100015    25437    23167    22945    22728    22516    22516  
1100023   107345   106168   107863   109523   111148   111148  
1100031     6224     5438     5312     5188     5067     5067  
1100049    88507    84813    85359    85893    86416    86416  
1100056    17934    16444    16323    16204    16088    16088  
...          ...      ...      ...      ...      ...      ...  
5222005    13675    13746    13863    13977    14088    14088  
5222054     8397     8611     8743     8873     9002     9002  
5222203     5731     6026     6171     6312     6451     6451  
5222302     5690     5758     5821     5882     5941     5941  
5300108  3039444  2974703  3015268  3055149  3094325  3094325  

[5570 rows x 13 columns]

In [5]:
# Checando quantidade de municípios
len(df_pop)

5570

In [6]:
# checando tipo de variáveis resultante da função 
df_pop.dtypes

2010    object
2011    object
2012    object
2013    object
2014    object
2015    object
2016    object
2017    object
2018    object
2019    object
2020    object
2021    object
2022    object
dtype: object

In [7]:
# convertendo dados para tipo integer
df_pop = df_pop.astype(int)

In [8]:
# Checando valores iguais a zero em cada coluna
num_zeros_por_coluna = (df_pop == 0).sum()
print(num_zeros_por_coluna)
del num_zeros_por_coluna # exclui a var temp, após verificação

2010    0
2011    0
2012    0
2013    0
2014    0
2015    0
2016    0
2017    0
2018    0
2019    0
2020    0
2021    0
2022    0
dtype: int64


Nota importante: cinco municípios dos 5.570 existentes atualmente não constavam nas bases de dados do IBGE entre 2010 e 2012, provavelmente foram criados após essa data. Os municípios tem os seguintes códigos: 1504752, 4212650, 4220000, 4314548, 5006275. Os dados inexistenets foram preenchidos com cálculo de população decrescente de 2012, para 2011 e para 2010 com abse na diferença da população anual dos anos subsequentes. Desse modo a matriz não resultará em dados nulos quando for utilizada para cálculos com outras matrizes para obtenção de dados relacionados à população, como é o caso da quantidade de procedimentos por 10 mil habitantes. 


In [9]:
import requests
# obtendo dados dos municípios listados por meio da API do IBGE

def obter_dados_municipios(codigos):
    dados_municipios = []
    for codigo in codigos.split("|"):
        try:
            url = f'https://servicodados.ibge.gov.br/api/v1/localidades/municipios/{codigo}'
            response = requests.get(url)
            if response.status_code == 200:
                dados_municipio = response.json()
                nome_municipio = dados_municipio['nome']
                uf_municipio = dados_municipio['microrregiao']['mesorregiao']['UF']['sigla']
                dados_municipios.append({'Nome': nome_municipio, 'UF': uf_municipio})
        except requests.exceptions.RequestException as e:
            print(f"Erro ao obter dados do município {codigo}: {e}")
    return dados_municipios



In [10]:
# fornece parâmetros com os códigos dos municípios para obtenção do nome e UF
codigos = '1504752|4212650|4220000|4314548|5006275'
municipios = obter_dados_municipios(codigos)
for municipio in municipios:
    print(f"Nome: {municipio['Nome']}, UF: {municipio['UF']}")


Nome: Mojuí dos Campos, UF: PA
Nome: Pescaria Brava, UF: SC
Nome: Balneário Rincão, UF: SC
Nome: Pinto Bandeira, UF: RS
Nome: Paraíso das Águas, UF: MS


3 criados em 2013, sendo que Balneário Rincão e Pinto Bandeira foram emancipados antes de 2010. De toda forma os dados de população para os 5 municípios nos 3 anos foram projetados de maneira a não causar viés nos dados.

In [11]:
# mostando os dados dos municípios e as projeções resultantes para população nos anos de 2010 a 2012. 
df_pop.loc[['1504752', '4212650', '4220000', '4314548', '5006275']]


2010   2011   2012   2013   2014   2015   2016   2017   2018   2019  \
codigo                                                                          
1504752  14905  15014  15123  15232  15341  15446  15548  15646  15982  16084   
4212650   9465   9539   9613   9687   9761   9835   9908   9980  10022  10091   
4220000  11040  11236  11432  11628  11824  12018  12212  12403  12570  12760   
4314548   2324   2443   2562   2681   2800   2824   2847   2868   2968   3003   
5006275   4627   4732   4837   4942   5047   5150   5251   5350   5455   5555   

          2020   2021   2022  
codigo                        
1504752  16184  16282  16282  
4212650  10159  10225  10225  
4220000  12946  13129  13129  
4314548   3036   3068   3068  
5006275   5654   5751   5751

### Etapa 2 - Criação do dataframe com dados dos procedimentos para calcular matrizes de treinamento da rede neural autoencoder mlp e dos algoritmos de ML supervisionado e não supervisionado

In [12]:
df_sia = pd.read_csv('./data/SIA_procedimentos.csv', sep='|')
df_sia

cnes  competencia                      estabelecimento  \
0            329       201105  CAPS ESTACAO CIDADANIA                
1            329       201102  CAPS ESTACAO CIDADANIA                
2            329       201111  CAPS ESTACAO CIDADANIA                
3            329       201209  CAPS ESTACAO CIDADANIA                
4            329       201104  CAPS ESTACAO CIDADANIA                
...          ...          ...                                  ...   
7477277  9993827       202009  FARMACIA CIDADA ESTADUAL DE GUACUI    
7477278  9993827       202110  FARMACIA CIDADA ESTADUAL DE GUACUI    
7477279  9993827       202208  FARMACIA CIDADA ESTADUAL DE GUACUI    
7477280  9993827       202207  FARMACIA CIDADA ESTADUAL DE GUACUI    
7477281  9995471       202011  HOFTALMED                             

                                razao_social  uf  codigo_municipio  \
0        PREFEITURA MUNICIPAL DO CABO DE SAN  PE          260290.0   
1        PREFEITURA MUNICIPAL DO CABO DE SAN  PE          260290.0   
2        PREFEITURA MUNICIPAL DO CABO DE SAN  PE          260290.0   
3        PREFEITURA MUNICIPAL DO CABO DE SAN  PE          260290.0   
4        PREFEITURA MUNICIPAL DO CABO DE SAN  PE          260290.0   
...                                      ...  ..               ...   
7477277  SECRETARIA DE ESTADO DA SAUDE        ES          320230.0   
7477278  SECRETARIA DE ESTADO DA SAUDE        ES          320230.0   
7477279  SECRETARIA DE ESTADO DA SAUDE        ES          320230.0   
7477280  SECRETARIA DE ESTADO DA SAUDE        ES          320230.0   
7477281  CRISTALINK CLINICA MEDICA LTDA       SC          420420.0   

                       municipio  co_procedimento  \
0        CABO DE SANTO AGOSTINHO         30108006   
1        CABO DE SANTO AGOSTINHO         30108010   
2        CABO DE SANTO AGOSTINHO         30108010   
3        CABO DE SANTO AGOSTINHO         30108010   
4        CABO DE SANTO AGOSTINHO         30108012   
...                          ...              ...   
7477277                   GUACUI         60480001   
7477278                   GUACUI         60483001   
7477279                   GUACUI         60483001   
7477280                   GUACUI         60483002   
7477281                  CHAPECO         40503013   

                                              procedimento  \
0        ACOMPANHAMENTO INTENSIVO DE PACIENTE EM SAUDE ...   
1        ACOMPANHAMENTO NAO INTENSIVO DE PACIENTE EM SA...   
2        ACOMPANHAMENTO NAO INTENSIVO DE PACIENTE EM SA...   
3        ACOMPANHAMENTO NAO INTENSIVO DE PACIENTE EM SA...   
4        ACOMPANHAMENTO SEMI-INTENSIVO DE PACIENTES EM ...   
...                                                    ...   
7477277  ENOXAPARINA SÓDICA 40 MG/0,4 ML SOLUÇÃO INJETÁ...   
7477278  SACUBITRIL VALSARTANA SÓDICA HIDRATADA 50 MG (...   
7477279  SACUBITRIL VALSARTANA SÓDICA HIDRATADA 50 MG (...   
7477280  SACUBITRIL VALSARTANA SÓDICA HIDRATADA 100 MG ...   
7477281                               VITRECTOMIA ANTERIOR   

         qtde_procedimentos  valor_procedimentos  
0                       166              3004.60  
1                        52               772.20  
2                        30               445.50  
3                        42               623.70  
4                       110              1749.00  
...                     ...                  ...  
7477277                  30                 0.00  
7477278                  56               193.76  
7477279                  56               193.76  
7477280                  60               207.60  
7477281                   1               381.08  

[7477282 rows x 11 columns]

In [13]:
df_sia.rename(columns={'co_procedimento': 'codigo_procedimento'}, inplace=True)

In [14]:
df_sih = pd.read_csv('./data/SIH_procedimentos.csv', sep='|')
df_sih

cnes  competencia                           estabelecimento  \
0              35       201002                    HOSPITAL MENDO SAMPAIO   
1              35       201002                    HOSPITAL MENDO SAMPAIO   
2              35       201003                    HOSPITAL MENDO SAMPAIO   
3              35       201003                    HOSPITAL MENDO SAMPAIO   
4              35       201004                    HOSPITAL MENDO SAMPAIO   
...           ...          ...                                       ...   
28657071  9950931       202211  HOSPITAL MUNICIPAL INTEGRADO SANTO AMARO   
28657072  9950931       202211  HOSPITAL MUNICIPAL INTEGRADO SANTO AMARO   
28657073  9994289       202105    POLICLINICA MUNICIPAL ESTEFANIA FARIAS   
28657074  9997784       202107      HOSPITAL E MATERNIDADE NAILA GONCALO   
28657075  9997784       202205      HOSPITAL E MATERNIDADE NAILA GONCALO   

                                             razao_social  uf  \
0         PREFEITURA MUNICIPAL DO CABO DE SANTO AGOSTINHO  PE   
1         PREFEITURA MUNICIPAL DO CABO DE SANTO AGOSTINHO  PE   
2         PREFEITURA MUNICIPAL DO CABO DE SANTO AGOSTINHO  PE   
3         PREFEITURA MUNICIPAL DO CABO DE SANTO AGOSTINHO  PE   
4         PREFEITURA MUNICIPAL DO CABO DE SANTO AGOSTINHO  PE   
...                                                   ...  ..   
28657071             PREFEITURA DO MUNICIPIO DE SAO PAULO  SP   
28657072             PREFEITURA DO MUNICIPIO DE SAO PAULO  SP   
28657073                  PREFEITURA MUNICIPAL DE SURUBIM  PE   
28657074                          MUNICIPIO DE BACABAEIRA  MA   
28657075                          MUNICIPIO DE BACABAEIRA  MA   

          codigo_municipio                municipio  codigo_procedimento  \
0                   260290  CABO DE SANTO AGOSTINHO            303080078   
1                   260290  CABO DE SANTO AGOSTINHO            303090316   
2                   260290  CABO DE SANTO AGOSTINHO            303020083   
3                   260290  CABO DE SANTO AGOSTINHO            409060135   
4                   260290  CABO DE SANTO AGOSTINHO            303070102   
...                    ...                      ...                  ...   
28657071            355030                SAO PAULO            407040099   
28657072            355030                SAO PAULO            407040102   
28657073            261450                  SURUBIM            303010223   
28657074            210125                BACABEIRA            411010034   
28657075            210125                BACABEIRA            310010039   

                                               procedimento  \
0                             TRATAMENTO DE ESTREPTOCOCCIAS   
1              TRATAMENTO DAS POLIARTROPATIAS INFLAMATORIAS   
2         TRATAMENTO DE OUTRAS DOENÇAS DO SANGUE E DOS Ó...   
3                                       HISTERECTOMIA TOTAL   
4         TRATAMENTO DE OUTRAS DOENCAS DO APARELHO DIGES...   
...                                                     ...   
28657071                 HERNIOPLASTIA INGUINAL (BILATERAL)   
28657072       HERNIOPLASTIA INGUINAL / CRURAL (UNILATERAL)   
28657073  TRATAMENTO DE INFECÇÃO PELO CORONAVIRUS – COVI...   
28657074                                    PARTO CESARIANO   
28657075                                       PARTO NORMAL   

          qtde_procedimentos  valor_procedimentos  
0                         11               340.62  
1                         11               238.70  
2                         24               388.37  
3                         12               658.03  
4                         15               380.58  
...                      ...                  ...  
28657071                  27              5490.54  
28657072                  51             10845.49  
28657073                  44             33000.00  
28657074                  74              4867.57  
28657075                  68              3731.10  

[28657076 rows x 11 colu

In [15]:
df_sia['fonte'] = 'sia'
df_sih['fonte'] = 'sih'

In [16]:
# Definindo a opção "float_format" para exibir somente 1 dígito após a vírgula
pd.set_option("float_format", "{:.1f}".format)

In [17]:
df_sia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7477282 entries, 0 to 7477281
Data columns (total 12 columns):
 #   Column               Dtype  
---  ------               -----  
 0   cnes                 int64  
 1   competencia          int64  
 2   estabelecimento      object 
 3   razao_social         object 
 4   uf                   object 
 5   codigo_municipio     float64
 6   municipio            object 
 7   codigo_procedimento  int64  
 8   procedimento         object 
 9   qtde_procedimentos   int64  
 10  valor_procedimentos  float64
 11  fonte                object 
dtypes: float64(2), int64(4), object(6)
memory usage: 684.6+ MB


In [18]:
df_sih.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28657076 entries, 0 to 28657075
Data columns (total 12 columns):
 #   Column               Dtype  
---  ------               -----  
 0   cnes                 int64  
 1   competencia          int64  
 2   estabelecimento      object 
 3   razao_social         object 
 4   uf                   object 
 5   codigo_municipio     int64  
 6   municipio            object 
 7   codigo_procedimento  int64  
 8   procedimento         object 
 9   qtde_procedimentos   int64  
 10  valor_procedimentos  float64
 11  fonte                object 
dtypes: float64(1), int64(5), object(6)
memory usage: 2.6+ GB


In [19]:
df_procedimentos = pd.concat([df_sia, df_sih])
df_procedimentos.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 36134358 entries, 0 to 28657075
Data columns (total 12 columns):
 #   Column               Dtype  
---  ------               -----  
 0   cnes                 int64  
 1   competencia          int64  
 2   estabelecimento      object 
 3   razao_social         object 
 4   uf                   object 
 5   codigo_municipio     float64
 6   municipio            object 
 7   codigo_procedimento  int64  
 8   procedimento         object 
 9   qtde_procedimentos   int64  
 10  valor_procedimentos  float64
 11  fonte                object 
dtypes: float64(2), int64(4), object(6)
memory usage: 3.5+ GB


In [20]:
# cria um df com base em uma pivot table para os dados de quantidade de procedimento por cnes, com colunas com dados do município
df_qtde_proc_cnes = (
    df_procedimentos
    .pivot_table(index=['codigo_municipio', 'municipio', 'uf', 'cnes', 'codigo_procedimento', 'procedimento'], columns='competencia', values='qtde_procedimentos')
    .reset_index()
    .sort_index()
)

In [21]:
# Convertendo o índice "date" em uma coluna
df_qtde_proc_cnes = df_qtde_proc_cnes.reset_index()

# Removendo o nome da coluna "variable" da primeira linha
df_qtde_proc_cnes = df_qtde_proc_cnes.rename_axis(None, axis=1)
df_qtde_proc_cnes  = df_qtde_proc_cnes.drop('index', axis =1)

In [22]:
# modifica tipo de variável para codigo do município para inteiro
df_qtde_proc_cnes['codigo_municipio'] = df_qtde_proc_cnes['codigo_municipio'].astype(int)
df_qtde_proc_cnes

codigo_municipio                            municipio  uf     cnes  \
0                  110001                ALTA FLORESTA D'OESTE  RO  2679477   
1                  110001                ALTA FLORESTA D'OESTE  RO  2679477   
2                  110001                ALTA FLORESTA D'OESTE  RO  2679477   
3                  110001                ALTA FLORESTA D'OESTE  RO  2679477   
4                  110001                ALTA FLORESTA D'OESTE  RO  2679477   
...                   ...                                  ...  ..      ...   
1377489            530180  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF  7988303   
1377490            530180  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF  7988303   
1377491            530180  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF  7988303   
1377492            530180  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF  7988303   
1377493            530180  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF  7988303   

         codigo_procedimento  \
0                  301060010   
1                  301060088   
2                  303010010   
3                  303010037   
4                  303010045   
...                      ...   
1377489             70101020   
1377490             70101021   
1377491             70101022   
1377492             70101024   
1377493             70101025   

                                              procedimento  201001  201002  \
0        DIAGNOSTICO E/OU ATENDIMENTO DE URGENCIA EM CL...     NaN     NaN   
1        DIAGNOSTICO E/OU ATENDIMENTO DE URGENCIA EM CL...     NaN     NaN   
2                            TRATAMENTO DE DENGUE CLÁSSICA   139.0   113.0   
3                 TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS     4.0     9.0   
4             TRATAMENTO DE DOENÇAS BACTERIANAS ZOONÓTICAS     NaN     NaN   
...                                                    ...     ...     ...   
1377489                         CADEIRA DE RODAS MONOBLOCO     NaN     NaN   
1377490                      CADEIRA DE RODAS (ACIMA 90KG)     NaN     NaN   
1377491     CADEIRA DE RODAS MOTORIZADA ADULTO OU INFANTIL     NaN     NaN   
1377492  CADEIRA DE RODAS PARA BANHO COM ENCOSTO RECLIN...     NaN     NaN   
1377493   CADEIRA DE RODAS PARA BANHO COM ARO DE PROPULSÃO     NaN     NaN   

         201003  201004  ...  202203  202204  202205  202206  202207  202208  \
0           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2          30.0    19.0  ...    12.0     6.0    28.0   138.0   261.0   189.0   
3          12.0     6.0  ...     4.0    22.0    12.0    56.0   157.0     NaN   
4           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
1377489     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
1377490     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
1377491     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
1377492     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
1377493     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

         202209  202210  202211  202212  
0           NaN    38.0    55.0     NaN  
1           NaN    82.0    94.0     NaN  
2          91.0   134.0   183.0     NaN  
3          63.0   249.0     2.0     NaN  
4           NaN     NaN     NaN     NaN  
...         ...     ...     ...     ...  
1377489     NaN     NaN     NaN     NaN  
1377490     NaN     NaN     NaN     NaN  
1377491     NaN     NaN     NaN     NaN  
1377492     NaN     NaN     NaN     NaN  
1377493     NaN     NaN     NaN     NaN  

[1377494 rows x 162 columns]

In [52]:
df_qtde_proc_cnes.to_csv('./data/df_qtde_proc_cnes.csv', sep = '|', encoding = 'UTF-8')

In [23]:
# verifica a média de valores NAN mas colunas...  que se mostra muito alto, pois se temos um df com 1,37 milhões de linhas e a média de linhas com valores NaN é de 1,1 milhões, 
# então precisamos selecionar somente os procedimentos com poucos valores ausentes nas linhas tanto para CNES quanto apra municípios
df_qtde_proc_cnes.isna().sum().mean()

1103507.2654320989

In [24]:
df_qtde_proc_mun = (
    df_procedimentos
    .pivot_table(index=['codigo_municipio', 'municipio', 'uf', 'codigo_procedimento', 'procedimento'], columns='competencia', values='qtde_procedimentos')
    .reset_index()
    .sort_index()
)

In [25]:
# Convertendo o índice "date" em uma coluna
df_qtde_proc_mun = df_qtde_proc_mun.reset_index()

# Removendo o nome da coluna "variable" da primeira linha
df_qtde_proc_mun = df_qtde_proc_mun.rename_axis(None, axis=1)
df_qtde_proc_mun  = df_qtde_proc_mun.drop('index', axis =1)
df_qtde_proc_mun

codigo_municipio                            municipio  uf  \
0               110001.0                ALTA FLORESTA D'OESTE  RO   
1               110001.0                ALTA FLORESTA D'OESTE  RO   
2               110001.0                ALTA FLORESTA D'OESTE  RO   
3               110001.0                ALTA FLORESTA D'OESTE  RO   
4               110001.0                ALTA FLORESTA D'OESTE  RO   
...                  ...                                  ...  ..   
932634          530180.0  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932635          530180.0  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932636          530180.0  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932637          530180.0  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   
932638          530180.0  REGIAO ADMINISTRATIVA DE TAGUATINGA  DF   

        codigo_procedimento  \
0                  30108012   
1                 301060010   
2                 301060088   
3                 303010010   
4                 303010037   
...                     ...   
932634            416120032   
932635            416120040   
932636            416120059   
932637            505010097   
932638            505010119   

                                             procedimento  201001  201002  \
0       ACOMPANHAMENTO SEMI-INTENSIVO DE PACIENTES EM ...     NaN   168.0   
1       DIAGNOSTICO E/OU ATENDIMENTO DE URGENCIA EM CL...     NaN     NaN   
2       DIAGNOSTICO E/OU ATENDIMENTO DE URGENCIA EM CL...     NaN     NaN   
3                           TRATAMENTO DE DENGUE CLÁSSICA   139.0   113.0   
4                TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS     4.0     9.0   
...                                                   ...     ...     ...   
932634                   MASTECTOMIA SIMPLES EM ONCOLOGIA     NaN     NaN   
932635  RESSECÇÃO DE LESÃO NÃO PALPÁVEL DE MAMA COM MA...     NaN     NaN   
932636  SEGMENTECTOMIA/QUADRANTECTOMIA/SETORECTOMIA DE...     NaN     NaN   
932637                              TRANSPLANTE DE CORNEA    54.0    48.0   
932638             TRANSPLANTE DE CORNEA (EM REOPERACOES)     NaN     NaN   

        201003  201004  201005  ...  202203  202204  202205  202206  202207  \
0        168.0   768.0   840.0  ...     NaN     NaN     NaN     NaN     NaN   
1          NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
2          NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
3         30.0    19.0     4.0  ...    12.0     6.0    28.0   138.0   261.0   
4         12.0     6.0    10.0  ...     4.0    22.0    12.0    56.0   157.0   
...        ...     ...     ...  ...     ...     ...     ...     ...     ...   
932634    47.0     5.0    27.0  ...     NaN     NaN     NaN     NaN     NaN   
932635     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
932636     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   
932637    72.0    33.0    18.0  ...     NaN     NaN     NaN     NaN     NaN   
932638     NaN     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN   

        202208  202209  202210  202211  202212  
0          NaN     NaN     NaN     NaN     NaN  
1          NaN     NaN    38.0    55.0     NaN  
2          NaN     NaN    82.0    94.0     NaN  
3        189.0    91.0   134.0   183.0     NaN  
4          NaN    63.0   249.0     2.0     NaN  
...        ...     ...     ...     ...     ...  
932634     NaN     NaN     NaN     NaN     NaN  
932635     NaN     NaN     NaN     NaN     NaN  
932636     NaN     NaN     NaN     NaN     NaN  
932637     NaN     NaN     NaN     NaN     NaN  
932638     NaN     NaN     NaN     NaN     NaN  

[932639 rows x 161 columns]

In [53]:
df_qtde_proc_mun.to_csv('./data/df_qtde_proc_mun.csv', sep = '|', encoding = 'UTF-8')

In [26]:
df_qtde_proc_mun.isna().sum().mean()

728710.354037267

In [27]:
# Define o número mínimo de valores não nulos necessários para manter uma linha
n_cols = len(df_qtde_proc_cnes.columns)
min_valid_values = int(n_cols * 0.9)

# Filtra o DataFrame para manter apenas as linhas com pelo menos 90% dos valores não nulos
procedimentos_cnes_sem_nan = df_qtde_proc_cnes.dropna(thresh=min_valid_values).reset_index()

# Mantém apenas as colunas 'codigo_procedimento' e 'procedimento' e remove duplicatas
procedimentos_cnes_sem_nan = procedimentos_cnes_sem_nan[['codigo_procedimento', 'procedimento']].drop_duplicates()


In [28]:
procedimentos_cnes_sem_nan

codigo_procedimento                                       procedimento
0                303010037           TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS
1                303010061      TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS
2                303010193  TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...
3                303060107                   TRATAMENTO DE CRISE HIPERTENSIVA
4                303070064  TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...
...                    ...                                                ...
54561            408060638  TRATAMENTO CIRÚRGICO DE LUXAÇÃO / FRATURA-LUXA...
54844            303080086                       TRATAMENTO DE FARMACODERMIAS
55480            408030321  ARTRODESE TORACO-LOMBO-SACRA POSTERIOR, SETE N...
55482            408050667  TRATAMENTO CIRÚRGICO DE LESÃO AGUDA CAPSULO-LI...
55483            408060018            ALONGAMENTO / ENCURTAMENTO MIOTENDINOSO

[921 rows x 2 columns]

In [29]:
procedimentos_cnes_validos = len(procedimentos_cnes_sem_nan)

In [49]:
procedimentos_cnes_sem_nan.to_excel('./data/procedimentos_cnes_iris.xlsx')

NOTA 2: isso implica que no total, seriam 921 os procedimentos POR CNES com no mínimo 90% de valores válidos (não nulos) nas linhas que poderiam ser analizados pela IRIS sem que seja necesário artifícios de tratamento e substituição de missings por dados sintéticos que causem viéses nos dados (substituição de missings por dados calculados, como média ou mediana. Sendo que, para a solução proposta conforme método de auditoria do MS, a mediana seria mais adequada para não repetir os viéses dos outliers de possíveis fraudes e porque é a medida utilizada nos processos de auditoria como no caso do Maranhão)

In [30]:
# Define o número mínimo de valores não nulos necessários para manter uma linha
n_cols = len(df_qtde_proc_mun.columns)
min_valid_values = int(n_cols * 0.9)

# Filtra o DataFrame para manter apenas as linhas com pelo menos 90% dos valores não nulos
procedimentos_mun_sem_nan = df_qtde_proc_mun.dropna(thresh=min_valid_values).reset_index()

# Mantém apenas as colunas 'codigo_procedimento' e 'procedimento' e remove duplicatas
procedimentos_mun_sem_nan = procedimentos_mun_sem_nan[['codigo_procedimento', 'procedimento']].drop_duplicates()

procedimentos_mun_sem_nan

codigo_procedimento                                       procedimento
0                303010037           TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS
1                303010061      TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS
2                303010193  TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...
3                303060107                   TRATAMENTO DE CRISE HIPERTENSIVA
4                303070064  TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...
...                    ...                                                ...
48522            409060267                 SALPINGOPLASTIA VIDEOLAPAROSCOPICA
48542            412030098                                       PLEUROSTOMIA
48733            408060409               RETIRADA DE TRAÇÃO TRANS-ESQUELÉTICA
52951            408030321  ARTRODESE TORACO-LOMBO-SACRA POSTERIOR, SETE N...
53018            413040097                                 PREPARO DE RETALHO

[1176 rows x 2 columns]

In [31]:
procedimentos_mun_validos = len(procedimentos_mun_sem_nan)
procedimentos_mun_validos

1176

In [48]:
procedimentos_mun_sem_nan.to_excel('./data/procedimentos_mun_iris.xlsx')

### df_proc_cnes é o dataframe com as quantidades de procedimentos POR CNES com menos de 10% de valores NaN nas linhas (ou seja, nos 13 anos ou 156 colunas de competencia mes/ano)
### Este df deve ser a base da qual calculamos todas as demais (df com as distâncias das medianas e assim por diante.. e os mesmos procedimentos devem ser feitos com os df com valores)

In [32]:
# Define o número mínimo de valores não nulos necessários para manter uma linha
n_cols = len(df_qtde_proc_cnes.columns)
min_valid_values = int(n_cols * 0.9)

# Filtra o DataFrame para manter apenas as linhas com pelo menos 90% dos valores não nulos
df_proc_cnes = df_qtde_proc_cnes.dropna(thresh=min_valid_values).reset_index()

df_proc_cnes = df_proc_cnes.drop('index', axis =1)

df_proc_cnes

codigo_municipio              municipio  uf     cnes  \
0                110001  ALTA FLORESTA D'OESTE  RO  2679477   
1                110001  ALTA FLORESTA D'OESTE  RO  2679477   
2                110001  ALTA FLORESTA D'OESTE  RO  2679477   
3                110001  ALTA FLORESTA D'OESTE  RO  2679477   
4                110001  ALTA FLORESTA D'OESTE  RO  2679477   
...                 ...                    ...  ..      ...   
55600            530010               BRASILIA  DF  3077098   
55601            530010               BRASILIA  DF  3077098   
55602            530010               BRASILIA  DF  3206874   
55603            530010               BRASILIA  DF  3206874   
55604            530010               BRASILIA  DF  3224228   

       codigo_procedimento                                       procedimento  \
0                303010037           TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS   
1                303010061      TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS   
2                303010193  TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...   
3                303060107                   TRATAMENTO DE CRISE HIPERTENSIVA   
4                303070064  TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...   
...                    ...                                                ...   
55600             70103013  APARELHO DE AMPLIFICAÇÃO SONORA INDIVIDUAL (AA...   
55601             70103014  APARELHO DE AMPLIFICAÇÃO SONORA INDIVIDUAL (AA...   
55602             50301001  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...   
55603             50401002                  PROCESSAMENTO DE CORNEA / ESCLERA   
55604             30501016  MANUTENÇÃO E ACOMPANHAMENTO DOMICILIAR DE PACI...   

       201001  201002  201003  201004  ...  202203  202204  202205  202206  \
0         4.0     9.0    12.0     6.0  ...     4.0    22.0    12.0    56.0   
1       116.0   109.0   137.0    89.0  ...    16.0    23.0    14.0    70.0   
2        12.0     NaN     4.0     7.0  ...    20.0    16.0     8.0    76.0   
3         9.0    11.0    37.0     4.0  ...    14.0    16.0    16.0    11.0   
4        10.0    18.0    10.0    12.0  ...    10.0     8.0    10.0     3.0   
...       ...     ...     ...     ...  ...     ...     ...     ...     ...   
55600    16.0    84.0    65.0    72.0  ...    63.0    63.0    31.0    63.0   
55601    14.0    36.0    92.0    30.0  ...    27.0    27.0    27.0    27.0   
55602     NaN    62.0    77.0    80.0  ...    62.0    82.0    70.0    99.0   
55603     NaN   129.0   170.0   198.0  ...   158.0   155.0   165.0   227.0   
55604     NaN     NaN     NaN     NaN  ...  2027.0  3045.0  1912.0  1949.0   

       202207  202208  202209  202210  202211  202212  
0       157.0     NaN    63.0   249.0     2.0     NaN  
1       105.0   108.0     6.0    45.0   163.0     NaN  
2       307.0    87.0   120.0    46.0   220.0     NaN  
3        91.0    27.0    50.0     8.0    37.0     NaN  
4        33.0   258.0     4.0    17.0    12.0     NaN  
...       ...     ...     ...     ...     ...     ...  
55600    63.0     NaN     NaN     NaN    13.0    38.0  
55601    27.0     NaN     NaN     NaN     6.0    19.0  
55602    53.0    62.0    29.0    66.0    40.0     NaN  
55603   119.0   142.0    42.0   150.0    88.0     NaN  
55604   864.0  2061.0  2071.0  3545.0  2028.0  1385.0  

[55605 rows x 162 columns]

### df_proc_mun é o dataframe com as quantidades de procedimentos POR MUNICÍPIO com menos de 10% de valores NaN nas linhas (ou seja, nos 13 anos ou 156 colunas de competencia mes/ano)
### Este df deve ser a base da qual calculamos todas as demais (df com as distâncias das medianas, df com a quantidade de procedimentos para cada 10 mil hab. e assim por diante.. e os mesmos procedimentos devem ser feitos com os df com valores)

In [33]:
# Define o número mínimo de valores não nulos necessários para manter uma linha
n_cols = len(df_qtde_proc_mun.columns)
min_valid_values = int(n_cols * 0.9)

# Filtra o DataFrame para manter apenas as linhas com pelo menos 90% dos valores não nulos
df_proc_mun = df_qtde_proc_mun.dropna(thresh=min_valid_values).reset_index()

# drop coluna desnecassária criada nos procedimetnos anteriores
df_proc_mun = df_proc_mun.drop('index', axis =1)

# modifica tipo de variável para codigo do município para inteiro
df_proc_mun['codigo_municipio'] = df_proc_mun['codigo_municipio'].astype(int)

df_proc_mun

codigo_municipio              municipio  uf  codigo_procedimento  \
0                110001  ALTA FLORESTA D'OESTE  RO            303010037   
1                110001  ALTA FLORESTA D'OESTE  RO            303010061   
2                110001  ALTA FLORESTA D'OESTE  RO            303010193   
3                110001  ALTA FLORESTA D'OESTE  RO            303060107   
4                110001  ALTA FLORESTA D'OESTE  RO            303070064   
...                 ...                    ...  ..                  ...   
53028            530010               BRASILIA  DF            416120024   
53029            530010               BRASILIA  DF            416120059   
53030            530010               BRASILIA  DF            503010014   
53031            530010               BRASILIA  DF            505010097   
53032            530010               BRASILIA  DF            506020045   

                                            procedimento  201001  201002  \
0               TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS     4.0     9.0   
1          TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS   116.0   109.0   
2      TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...    12.0     NaN   
3                       TRATAMENTO DE CRISE HIPERTENSIVA     9.0    11.0   
4      TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...    10.0    18.0   
...                                                  ...     ...     ...   
53028  MASTECTOMIA RADICAL COM LINFADENECTOMIA AXILAR...    87.5    29.0   
53029  SEGMENTECTOMIA/QUADRANTECTOMIA/SETORECTOMIA DE...     6.0     NaN   
53030  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...    16.0   184.0   
53031                              TRANSPLANTE DE CORNEA    31.0     9.0   
53032  TRATAMENTO DE INTERCORRÊNCIA PÓS-TRANSPLANTE D...   955.0   560.5   

       201003  201004  201005  ...  202203  202204  202205  202206  202207  \
0        12.0     6.0    10.0  ...     4.0    22.0    12.0    56.0   157.0   
1       137.0    89.0    74.0  ...    16.0    23.0    14.0    70.0   105.0   
2         4.0     7.0     9.0  ...    20.0    16.0     8.0    76.0   307.0   
3        37.0     4.0    19.0  ...    14.0    16.0    16.0    11.0    91.0   
4        10.0    12.0     8.0  ...    10.0     8.0    10.0     3.0    33.0   
...       ...     ...     ...  ...     ...     ...     ...     ...     ...   
53028    16.0    46.3    49.5  ...    34.0    43.2    19.8    17.8    27.5   
53029    56.5     4.0    23.0  ...    76.6    34.0    55.5    40.4    29.8   
53030    13.0    26.0    57.0  ...    78.0    69.0    27.2   117.7    53.0   
53031    47.5    20.5     NaN  ...    67.5   116.0    59.5    65.5    55.0   
53032   986.0  1023.5   570.5  ...  1664.8  3103.3  2100.8  2082.7  2336.5   

       202208  202209  202210  202211  202212  
0         NaN    63.0   249.0     2.0     NaN  
1       108.0     6.0    45.0   163.0     NaN  
2        87.0   120.0    46.0   220.0     NaN  
3        27.0    50.0     8.0    37.0     NaN  
4       258.0     4.0    17.0    12.0     NaN  
...       ...     ...     ...     ...     ...  
53028    16.0    27.8   123.3    35.0    47.8  
53029    64.6    48.8   123.0    59.5    47.7  
53030   149.5    70.7    52.0    81.0    26.7  
53031    29.0    39.0    84.5   145.5     5.0  
53032  1707.2   630.7   915.2  1520.7   478.0  

[53033 rows x 161 columns]

In [34]:
# Selecionar apenas colunas numéricas após a quarta coluna
num_cols = df_proc_mun.columns[5:]

# Substituindo NaN pela mediana de cada linha
df_proc_mun[num_cols] = df_proc_mun[num_cols].apply(lambda x: x.fillna(x.median()), axis=1)

# Calculando o desvio da mediana para cada célula
desvios_med_n_mun = df_proc_mun[num_cols].apply(lambda x: x - x.median(), axis=1)

# Criando um novo DataFrame com os desvios_med_n como uma matriz (m)
m_desvios_med_n_mun = pd.DataFrame(desvios_med_n_mun.values, index=df_proc_mun.index, columns=num_cols)

# Concatenando as quatro primeiras colunas do DataFrame original com o novo DataFrame de desvios_med_n_mun
df_desv_med_n_proc_mun = pd.concat([df_proc_mun.iloc[:, :5], m_desvios_med_n_mun], axis=1)
df_desv_med_n_proc_mun

codigo_municipio              municipio  uf  codigo_procedimento  \
0                110001  ALTA FLORESTA D'OESTE  RO            303010037   
1                110001  ALTA FLORESTA D'OESTE  RO            303010061   
2                110001  ALTA FLORESTA D'OESTE  RO            303010193   
3                110001  ALTA FLORESTA D'OESTE  RO            303060107   
4                110001  ALTA FLORESTA D'OESTE  RO            303070064   
...                 ...                    ...  ..                  ...   
53028            530010               BRASILIA  DF            416120024   
53029            530010               BRASILIA  DF            416120059   
53030            530010               BRASILIA  DF            503010014   
53031            530010               BRASILIA  DF            505010097   
53032            530010               BRASILIA  DF            506020045   

                                            procedimento  201001  201002  \
0               TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS    -4.0     1.0   
1          TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS    79.0    72.0   
2      TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...    -5.0     0.0   
3                       TRATAMENTO DE CRISE HIPERTENSIVA   -11.0    -9.0   
4      TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...    -4.0     4.0   
...                                                  ...     ...     ...   
53028  MASTECTOMIA RADICAL COM LINFADENECTOMIA AXILAR...    37.5   -21.0   
53029  SEGMENTECTOMIA/QUADRANTECTOMIA/SETORECTOMIA DE...   -29.0     0.0   
53030  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...   -86.8    81.2   
53031                              TRANSPLANTE DE CORNEA   -20.0   -42.0   
53032  TRATAMENTO DE INTERCORRÊNCIA PÓS-TRANSPLANTE D...  -169.8  -564.3   

       201003  201004  201005  ...  202203  202204  202205  202206  202207  \
0         4.0    -2.0     2.0  ...    -4.0    14.0     4.0    48.0   149.0   
1       100.0    52.0    37.0  ...   -21.0   -14.0   -23.0    33.0    68.0   
2       -13.0   -10.0    -8.0  ...     3.0    -1.0    -9.0    59.0   290.0   
3        17.0   -16.0    -1.0  ...    -6.0    -4.0    -4.0    -9.0    71.0   
4        -4.0    -2.0    -6.0  ...    -4.0    -6.0    -4.0   -11.0    19.0   
...       ...     ...     ...  ...     ...     ...     ...     ...     ...   
53028   -34.0    -3.7    -0.5  ...   -16.0    -6.8   -30.2   -32.2   -22.5   
53029    21.5   -31.0   -12.0  ...    41.6    -1.0    20.5     5.4    -5.2   
53030   -89.8   -76.8   -45.8  ...   -24.8   -33.8   -75.6    14.8   -49.8   
53031    -3.5   -30.5     0.0  ...    16.5    65.0     8.5    14.5     4.0   
53032  -138.8  -101.3  -554.3  ...   539.9  1978.5   975.9   957.8  1211.7   

       202208  202209  202210  202211  202212  
0         0.0    55.0   241.0    -6.0     0.0  
1        71.0   -31.0     8.0   126.0     0.0  
2        70.0   103.0    29.0   203.0     0.0  
3         7.0    30.0   -12.0    17.0     0.0  
4       244.0   -10.0     3.0    -2.0     0.0  
...       ...     ...     ...     ...     ...  
53028   -34.0   -22.2    73.3   -15.0    -2.2  
53029    29.6    13.8    88.0    24.5    12.7  
53030    46.7   -32.2   -50.8   -21.8   -76.2  
53031   -22.0   -12.0    33.5    94.5   -46.0  
53032   582.4  -494.2  -209.6   395.8  -646.8  

[53033 rows x 161 columns]

In [47]:
df_desv_med_n_proc_mun.to_excel('./data/df_desv_med_n_proc_mun.xlsx')

In [42]:
df_proc_mun[num_cols] = df_proc_mun[num_cols].apply(lambda x: x.fillna(x.median()), axis=1)
df_proc_mun

codigo_municipio              municipio  uf  codigo_procedimento  \
0                110001  ALTA FLORESTA D'OESTE  RO            303010037   
1                110001  ALTA FLORESTA D'OESTE  RO            303010061   
2                110001  ALTA FLORESTA D'OESTE  RO            303010193   
3                110001  ALTA FLORESTA D'OESTE  RO            303060107   
4                110001  ALTA FLORESTA D'OESTE  RO            303070064   
...                 ...                    ...  ..                  ...   
53028            530010               BRASILIA  DF            416120024   
53029            530010               BRASILIA  DF            416120059   
53030            530010               BRASILIA  DF            503010014   
53031            530010               BRASILIA  DF            505010097   
53032            530010               BRASILIA  DF            506020045   

                                            procedimento  201001  201002  \
0               TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS     4.0     9.0   
1          TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS   116.0   109.0   
2      TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...    12.0    17.0   
3                       TRATAMENTO DE CRISE HIPERTENSIVA     9.0    11.0   
4      TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...    10.0    18.0   
...                                                  ...     ...     ...   
53028  MASTECTOMIA RADICAL COM LINFADENECTOMIA AXILAR...    87.5    29.0   
53029  SEGMENTECTOMIA/QUADRANTECTOMIA/SETORECTOMIA DE...     6.0    35.0   
53030  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...    16.0   184.0   
53031                              TRANSPLANTE DE CORNEA    31.0     9.0   
53032  TRATAMENTO DE INTERCORRÊNCIA PÓS-TRANSPLANTE D...   955.0   560.5   

       201003  201004  201005  ...  202203  202204  202205  202206  202207  \
0        12.0     6.0    10.0  ...     4.0    22.0    12.0    56.0   157.0   
1       137.0    89.0    74.0  ...    16.0    23.0    14.0    70.0   105.0   
2         4.0     7.0     9.0  ...    20.0    16.0     8.0    76.0   307.0   
3        37.0     4.0    19.0  ...    14.0    16.0    16.0    11.0    91.0   
4        10.0    12.0     8.0  ...    10.0     8.0    10.0     3.0    33.0   
...       ...     ...     ...  ...     ...     ...     ...     ...     ...   
53028    16.0    46.3    49.5  ...    34.0    43.2    19.8    17.8    27.5   
53029    56.5     4.0    23.0  ...    76.6    34.0    55.5    40.4    29.8   
53030    13.0    26.0    57.0  ...    78.0    69.0    27.2   117.7    53.0   
53031    47.5    20.5    51.0  ...    67.5   116.0    59.5    65.5    55.0   
53032   986.0  1023.5   570.5  ...  1664.8  3103.3  2100.8  2082.7  2336.5   

       202208  202209  202210  202211  202212  
0         8.0    63.0   249.0     2.0     8.0  
1       108.0     6.0    45.0   163.0    37.0  
2        87.0   120.0    46.0   220.0    17.0  
3        27.0    50.0     8.0    37.0    20.0  
4       258.0     4.0    17.0    12.0    14.0  
...       ...     ...     ...     ...     ...  
53028    16.0    27.8   123.3    35.0    47.8  
53029    64.6    48.8   123.0    59.5    47.7  
53030   149.5    70.7    52.0    81.0    26.7  
53031    29.0    39.0    84.5   145.5     5.0  
53032  1707.2   630.7   915.2  1520.7   478.0  

[53033 rows x 161 columns]

In [43]:
df_proc_mun.to_excel('./data/df_proc_mun.xlsx')

In [35]:
# Selecionar apenas colunas numéricas após a quarta coluna
num_cols = df_proc_cnes.columns[6:]

# Substituindo NaN pela mediana de cada linha
df_proc_cnes[num_cols] = df_proc_cnes[num_cols].apply(lambda x: x.fillna(x.median()), axis=1)

# Calculando o desvio da mediana para cada célula
desvios_med_n_cnes = df_proc_cnes[num_cols].apply(lambda x: x - x.median(), axis=1)

# Criando um novo DataFrame com os desvios_med_n como uma matriz (m)
m_desvios_med_n_cnes = pd.DataFrame(desvios_med_n_cnes.values, index=df_proc_cnes.index, columns=num_cols)

# Concatenando as quatro primeiras colunas do DataFrame original com o novo DataFrame de desvios_med_n_cnes
df_desv_med_n_proc_cnes = pd.concat([df_proc_cnes.iloc[:, :5], m_desvios_med_n_cnes], axis=1)
df_desv_med_n_proc_cnes

codigo_municipio              municipio  uf     cnes  \
0                110001  ALTA FLORESTA D'OESTE  RO  2679477   
1                110001  ALTA FLORESTA D'OESTE  RO  2679477   
2                110001  ALTA FLORESTA D'OESTE  RO  2679477   
3                110001  ALTA FLORESTA D'OESTE  RO  2679477   
4                110001  ALTA FLORESTA D'OESTE  RO  2679477   
...                 ...                    ...  ..      ...   
55600            530010               BRASILIA  DF  3077098   
55601            530010               BRASILIA  DF  3077098   
55602            530010               BRASILIA  DF  3206874   
55603            530010               BRASILIA  DF  3206874   
55604            530010               BRASILIA  DF  3224228   

       codigo_procedimento  201001  201002  201003  201004  201005  ...  \
0                303010037    -4.0     1.0     4.0    -2.0     2.0  ...   
1                303010061    79.0    72.0   100.0    52.0    37.0  ...   
2                303010193    -5.0     0.0   -13.0   -10.0    -8.0  ...   
3                303060107   -11.0    -9.0    17.0   -16.0    -1.0  ...   
4                303070064    -4.0     4.0    -4.0    -2.0    -6.0  ...   
...                    ...     ...     ...     ...     ...     ...  ...   
55600             70103013   -47.0    21.0     2.0     9.0    28.0  ...   
55601             70103014   -13.0     9.0    65.0     3.0    15.0  ...   
55602             50301001     0.0   -15.0     0.0     3.0     0.0  ...   
55603             50401002     0.0   -25.0    16.0    44.0    36.0  ...   
55604             30501016     0.0     0.0     0.0     0.0     0.0  ...   

       202203  202204  202205  202206  202207  202208  202209  202210  202211  \
0        -4.0    14.0     4.0    48.0   149.0     0.0    55.0   241.0    -6.0   
1       -21.0   -14.0   -23.0    33.0    68.0    71.0   -31.0     8.0   126.0   
2         3.0    -1.0    -9.0    59.0   290.0    70.0   103.0    29.0   203.0   
3        -6.0    -4.0    -4.0    -9.0    71.0     7.0    30.0   -12.0    17.0   
4        -4.0    -6.0    -4.0   -11.0    19.0   244.0   -10.0     3.0    -2.0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
55600     0.0     0.0   -32.0     0.0     0.0     0.0     0.0     0.0   -50.0   
55601     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   -21.0   
55602   -15.0     5.0    -7.0    22.0   -24.0   -15.0   -48.0   -11.0   -37.0   
55603     4.0     1.0    11.0    73.0   -35.0   -12.0  -112.0    -4.0   -66.0   
55604   440.0  1458.0   325.0   362.0  -723.0   474.0   484.0  1958.0   441.0   

       202212  
0         0.0  
1         0.0  
2         0.0  
3         0.0  
4         0.0  
...       ...  
55600   -25.0  
55601    -8.0  
55602     0.0  
55603     0.0  
55604  -202.0  

[55605 rows x 161 columns]

In [46]:
df_desv_med_n_proc_cnes.to_excel('./data/df_desv_med_n_proc_cnes.xlsx')

In [44]:
df_proc_cnes[num_cols] = df_proc_cnes[num_cols].apply(lambda x: x.fillna(x.median()), axis=1)
df_proc_cnes

codigo_municipio              municipio  uf     cnes  \
0                110001  ALTA FLORESTA D'OESTE  RO  2679477   
1                110001  ALTA FLORESTA D'OESTE  RO  2679477   
2                110001  ALTA FLORESTA D'OESTE  RO  2679477   
3                110001  ALTA FLORESTA D'OESTE  RO  2679477   
4                110001  ALTA FLORESTA D'OESTE  RO  2679477   
...                 ...                    ...  ..      ...   
55600            530010               BRASILIA  DF  3077098   
55601            530010               BRASILIA  DF  3077098   
55602            530010               BRASILIA  DF  3206874   
55603            530010               BRASILIA  DF  3206874   
55604            530010               BRASILIA  DF  3224228   

       codigo_procedimento                                       procedimento  \
0                303010037           TRATAMENTO DE OUTRAS DOENÇAS BACTERIANAS   
1                303010061      TRATAMENTO DE DOENÇAS INFECCIOSAS INTESTINAIS   
2                303010193  TRATAMENTO DE OUTRAS DOENÇAS CAUSADAS POR VÍRU...   
3                303060107                   TRATAMENTO DE CRISE HIPERTENSIVA   
4                303070064  TRATAMENTO DE DOENCAS DO ESOFAGO ESTOMAGO E DU...   
...                    ...                                                ...   
55600             70103013  APARELHO DE AMPLIFICAÇÃO SONORA INDIVIDUAL (AA...   
55601             70103014  APARELHO DE AMPLIFICAÇÃO SONORA INDIVIDUAL (AA...   
55602             50301001  AÇÕES RELACIONADAS A DOAÇÃO DE ÓRGÃOS E TECIDO...   
55603             50401002                  PROCESSAMENTO DE CORNEA / ESCLERA   
55604             30501016  MANUTENÇÃO E ACOMPANHAMENTO DOMICILIAR DE PACI...   

       201001  201002  201003  201004  ...  202203  202204  202205  202206  \
0         4.0     9.0    12.0     6.0  ...     4.0    22.0    12.0    56.0   
1       116.0   109.0   137.0    89.0  ...    16.0    23.0    14.0    70.0   
2        12.0    17.0     4.0     7.0  ...    20.0    16.0     8.0    76.0   
3         9.0    11.0    37.0     4.0  ...    14.0    16.0    16.0    11.0   
4        10.0    18.0    10.0    12.0  ...    10.0     8.0    10.0     3.0   
...       ...     ...     ...     ...  ...     ...     ...     ...     ...   
55600    16.0    84.0    65.0    72.0  ...    63.0    63.0    31.0    63.0   
55601    14.0    36.0    92.0    30.0  ...    27.0    27.0    27.0    27.0   
55602    77.0    62.0    77.0    80.0  ...    62.0    82.0    70.0    99.0   
55603   154.0   129.0   170.0   198.0  ...   158.0   155.0   165.0   227.0   
55604  1587.0  1587.0  1587.0  1587.0  ...  2027.0  3045.0  1912.0  1949.0   

       202207  202208  202209  202210  202211  202212  
0       157.0     8.0    63.0   249.0     2.0     8.0  
1       105.0   108.0     6.0    45.0   163.0    37.0  
2       307.0    87.0   120.0    46.0   220.0    17.0  
3        91.0    27.0    50.0     8.0    37.0    20.0  
4        33.0   258.0     4.0    17.0    12.0    14.0  
...       ...     ...     ...     ...     ...     ...  
55600    63.0    63.0    63.0    63.0    13.0    38.0  
55601    27.0    27.0    27.0    27.0     6.0    19.0  
55602    53.0    62.0    29.0    66.0    40.0    77.0  
55603   119.0   142.0    42.0   150.0    88.0   154.0  
55604   864.0  2061.0  2071.0  3545.0  2028.0  1385.0  

[55605 rows x 162 columns]

In [45]:
df_proc_cnes.to_excel('./data/df_proc_cnes.xlsx')

### Calculando o df_proc_pop com os dados de quantidade de procedimentos para cada 10 mil habitantes para os 13 anos. 

In [36]:
df_proc_mun.to_excel('./data/df_proc_mun.xlsx')

In [41]:
# Convertendo o índice "date" em uma coluna
#df_pop = df_pop.reset_index()
# drop coluna desnecassária criada nos procedimetnos anteriores
#df_pop = df_pop.drop('index', axis =1)
df_pop

codigo     2010     2011     2012     2013     2014     2015     2016  \
0     1100015    24392    24228    24069    25728    25652    25578    25506   
1     1100023    90353    91570    92747   101269   102860   104401   105896   
2     1100031     6313     6221     6132     6495     6424     6355     6289   
3     1100049    78574    78958    79330    85863    86556    87226    87877   
4     1100056    17029    16939    16852    18041    18013    17986    17959   
...       ...      ...      ...      ...      ...      ...      ...      ...   
5565  5222005    12548    12644    12737    13227    13343    13456    13567   
5566  5222054     7371     7475     7576     7933     8053     8171     8286   
5567  5222203     4735     4846     4954     5246     5371     5495     5615   
5568  5222302     5145     5195     5244     5460     5520     5578     5635   
5569  5300108  2570160  2609997  2648532  2789761  2852372  2914830  2977216   

         2017     2018     2019     2020     2021     2022  
0       25437    23167    22945    22728    22516    22516  
1      107345   106168   107863   109523   111148   111148  
2        6224     5438     5312     5188     5067     5067  
3       88507    84813    85359    85893    86416    86416  
4       17934    16444    16323    16204    16088    16088  
...       ...      ...      ...      ...      ...      ...  
5565    13675    13746    13863    13977    14088    14088  
5566     8397     8611     8743     8873     9002     9002  
5567     5731     6026     6171     6312     6451     6451  
5568     5690     5758     5821     5882     5941     5941  
5569  3039444  2974703  3015268  3055149  3094325  3094325  

[5570 rows x 14 columns]